Kernel/OrderingUtils.hpp

```diff
@@ -0,0 +1,501 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+
+#ifndef __OrderingUtils__
+#define __OrderingUtils__
+
+#include "Debug/Assertion.hpp"
+#include "Lib/Option.hpp"
+#include "Kernel/Theory.hpp"
+#include "Kernel/Ordering.hpp"
+#include "Lib/STL.hpp"
+#include "Lib/VirtualIterator.hpp"
+#include "Lib/Output.hpp"
+
+namespace Kernel {
+  using namespace Lib;
+
+  template<class T>
+  class MultiSet {
+    Stack<std::tuple<T, IntegerConstantType>> _elems;
+  public:
+    void integrity() const {
+#if VDEBUG
+      ASS(std::is_sorted(_elems.begin(), _elems.end(), [](auto l, auto r) { return std::get<0>(l) < std::get<0>(r); }))
+      for (auto e : _elems) {
+        ASS_G(std::get<1>(e), IntegerConstantType(0))
+      }
+#endif
+    }
```

Comment:
Did you consider implementing this in terms of `std::multiset`?

Reply:
Good point. I wasn't aware this existed tbh. I'll give it a think for future optimizations! I think changing the current impl to that would be quite a bit of effort, but it might also be able to give some nice performance benefits, so I'll give it a think :)